schema-registry/rest-service/src/main/java/com/hortonworks/registries/schemaregistry/webservice/ConfluentSchemaRegistryCompatibleResource.java

```diff
@@ -201,18 +201,42 @@ public Response getSchemaVersion(@ApiParam(value = "subject", required = true)
 
-            SchemaVersionInfo schemaVersionInfo = "latest".equals(versionId)
-                    ? schemaRegistry.getLatestSchemaVersionInfo(subject)
-                    : schemaRegistry.getSchemaVersionInfo(Long.valueOf(versionId));
+            if ("latest".equals(versionId)) {
+                schemaVersionInfo = schemaRegistry.getLatestSchemaVersionInfo(subject);
+            } else {
+                SchemaMetadataInfo schemaMetadataInfo = schemaRegistry.getSchemaMetadataInfo(subject);
+                if(schemaMetadataInfo == null) {
+                    throw new SchemaNotFoundException();
+                }
+                SchemaVersionInfo fetchedSchemaVersionInfo = null;
+                try {
+                    fetchedSchemaVersionInfo = schemaRegistry.getSchemaVersionInfo(Long.valueOf(versionId));
+                } catch (SchemaNotFoundException e) {
+                    LOG.error("Schema version not found with version id [{}]", versionId, e);
+                }
+                if (fetchedSchemaVersionInfo != null) {
+                    if(subject.equals(fetchedSchemaVersionInfo.getName())) {
+                        schemaVersionInfo = fetchedSchemaVersionInfo;
+                    } else {
+                        LOG.error("Received schema version for id [{}] belongs to subject [{}] which is different from requested subject [{}]",
+                                  versionId, fetchedSchemaVersionInfo.getName(), subject);
+                    }
+                }
+            }
 
-            SchemaVersionEntry schemaVersionEntry = new SchemaVersionEntry(schemaVersionInfo.getName(),
-                                                                           schemaVersionInfo.getId().intValue(),
-                                                                           schemaVersionInfo.getSchemaText());
-            response = WSUtils.respondEntity(schemaVersionEntry, Response.Status.OK);
+            if(schemaVersionInfo == null) {
+                response = versionNotFoundError();
+            } else {
+                SchemaVersionEntry schemaVersionEntry = new SchemaVersionEntry(schemaVersionInfo.getName(),
+                                                                               schemaVersionInfo.getId().intValue(),
+                                                                               schemaVersionInfo.getSchemaText());
+                response = WSUtils.respondEntity(schemaVersionEntry, Response.Status.OK);
+            }
         } catch (SchemaNotFoundException ex) {
-            LOG.error("No schema version found with id [{}]", versionId, ex);
-            response = versionNotFoundError();
+            LOG.error("No schema found with subject [{}]", subject, ex);
+            response = subjectNotFoundError();
         } catch (Exception ex) {
             LOG.error("Encountered error while retrieving all subjects", ex);
             response = serverError();
```

Comment:
No plans to address these as they are specific errors form storage etc. Users would anyway see the error message on the reason of failure of the invocation in their error message. Idea of this PR is to handle respective errors for no schema/version found.